### ISSUE TITLE
Missing dispose in ExecuteQuery

### ISSUE THREAD

**Issue by alkampfergit** [NEventStore/NEventStore.Persistence.SQL#20]
If you look at the code of method [SqlPersistenceEngine::ExecuteQuery](https://github.com/NEventStore/NEventStore.Persistence.SQL/blob/accc7c627f4ad093e0dcd601043fc4eb511d95bd/src/NEventStore.Persistence.Sql/SqlPersistenceEngine.cs#L374) you can notice that nor connection nor transaction gets disposed if the code is successful. 

Compare ExecuteQuery with the implementation of [SqlPersistenEngine::ExecuteCommand](https://github.com/NEventStore/NEventStore.Persistence.SQL/blob/accc7c627f4ad093e0dcd601043fc4eb511d95bd/src/NEventStore.Persistence.Sql/SqlPersistenceEngine.cs#L443)  that correctly disposes everything.

This can lead to serious problem, and actually probably it did not generated error because the standard implementation of OpenTransaction returns null. 

I think that all the disposable object should be wrapped in using clause.

**Comment by fschmied** [comment]
I think in the success case, the `PagedEnumerationCollection` and the `CommonDbStatement.Dispose` methods are responsible for disposing the connection and transaction after enumeration finished. As currently implemented, `SqlPersistenceEngine.ExecuteQuery` cannot dispose them in the success case (as enumeration is controlled by the caller).

However, this is really not intuitive and quite difficult to reason about, so I think it should be refactored. For example, the `SqlPersistenceEngine.ExecuteQuery` method could use an iterator block (`foreach` and `yield return`) wrapped in a `using` statement. That way, the `Dispose` methods would be executed whenever and however iteration finished.

**Comment by alkampfergit** [comment]
I agree with the refactoring, I've checked the code and all object seems to be correctly disposed by all code that calls that function, but the code is quite hard to read.
